main.py

```diff
@@ -168,31 +188,41 @@ def insert_center_and_mask(img: np.ndarray, final_image: (int, int)) -> (np.ndar
 
     return new_img, mask, zoom_factor_start, zoom_factor_end
 
+def get_upscale_image_max_width(image: Image) -> int:
+    '''Return the maximum upscale width that does not exceed the API's output image pixel limit'''
+    UPSCALE_IMAGE_OUTPUT_LIMIT = 4194304  # Defined in https://platform.stability.ai/docs/features/image-upscaling
 
-def generate_images(prompt, iterations, scaling, size, api_key, temp_dir=Path("output")):
+    im = cv2.imread(str(image.path))
+    h, w, _ = im.shape
+
+    upscale_width = 2560                  # Start with this width optimistically and scale down
+    upscale_height = (upscale_width * h) / w
+
+    while upscale_width * upscale_height > UPSCALE_IMAGE_OUTPUT_LIMIT:
```

Comment:
If you know the image's current height, width, and max size, then you can calculate the exact max width using some algebra. 

You don't need this guess loop.